feat: kcov

tkhr-sait opened the pull request:
**What kind of change does this PR introduce?**
- [x] feature
- [ ] bug fix
- [ ] docs update
- [ ] other

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our [guidelines](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#commits)
- [ ] Fix/Feature: Docs have been added/updated
- [x] Fix/Feature: Tests have been added; existing tests pass

**Does this PR introduce a breaking change?**
- [ ] yes
- [x] no


**What is the current behavior? (You can also link to an open issue here)**


**What is the new behavior (if this is a feature change)?**
kcov coverage

**Other information**:

tkhr-sait committed:
feat: kcov

tkhr-sait commented:
一応、こちらも。
shellcheck は stfw/test 下配置の想定で大丈夫ですかね（配布物には含めない）

tkhr-sait committed:
Merge branch 'master' of https://github.com/scenario-test-framework/stfw into feature/kcov

suwa-sh commented:
そうですね！

/test/ :ut
  ut_all.sh
  lib/stfw/adapter/
的に同じpkg構成で掘る感じかと
考えてますー

tkhr-sait commented:
了解でーす
X shellcheck
O shunitでした

tkhr-sait committed:
chore: shunit2 support

tkhr-sait commented:
一応、shunitのカバレジ取得まで
stfw/build/product/coverage.sh

tkhr-sait committed:
fix: 同一コードとしてマージされるよう修正

suwa-sh committed:
chore(build/product/coverage): 動作確認の結果を反映

kcov実行時の標準出力で、awkスクリプトは1行しか除外してくれない様子だったので、lib/commons/logging側を1lineに調整
build/product/coverage.sh を実行すると2つずつcontainerが残ってしまうので、削除するよう変更
実行結果の表示を追加

suwa-sh commented:
ちょろっと動作確認の結果を載せてみましたー
ワークフロー周りで気になったところがあったので、一旦整理しませんか？

1. **coverageを採るタイミングはいつにする？**
analyzeとセットで ut_all で実行。しきい値を用意して、下回ってたらエラーとか？
ut_allを実行するとcoverageも取られる or オプションで切り替えられる、とかですかね。。。

2. **ITのカバレッジは取る必要あるか？**
上のタイミングで実施するなら、IT時は不要かな、と。digdag経由が取れないですしねf(^_^;
動作確認＆UTを用意する優先順位付けにはありですが、UTを整えるならいらなくなるかな、と。

3. **test失敗時の制御を追加する必要がありそう**
kcov containerの動きが、まだ細かく見れてませんがテストスクリプトがエラー終了した時
どんな挙動にするかも決めておかないとですね〜

4. **travisから実行する場合、coveralls.ioに連携したい**
coveralls.ioに連携したいなーと思ってたんですが、kcovに機能あるんですね！
ワークフローが決まったらCIで実行する場合は、バッジ表示まで含めて連携できるようにしたいですねー

suwa-sh commented on src/bin/lib/commons/logging:
```diff
@@ -200,23 +200,7 @@ function private.log.format() {
-      print " "$0
-    }
-  '                                                                                                |
+    ' { printf _timestamp" "_log_level; for (i = 0; i < LOG__INDENT_COUNT; i++) { if ( i == 0) { printf " " }; printf LOG__INDENT_STR; }; print " "$0; }' |
```
メンテナンスしづらいので戻しますね。
kcov側の出力を制御するように考えてみますー

suwa-sh commented on test/ut/lib/stfw/domain/service/spec/process_spec_test.sh:
```diff
@@ -0,0 +1,28 @@
+
+# テスト対象の関数名 . はだめ _ は ok
+# bashの?で落ちる場合、処理続行されない（検証できない）
+function test_stfw_domain_service_spec_process_dirname() {
```
ネームスペースは外して、test_dirnname とかにしましょうか。
多分、test_dirname_ケース名 みたいな感じで増えていくと思うので、短めに抑えておいたほうが良いかなと。

suwa-sh commented on test/ut/lib/stfw/domain/service/spec/process_spec_test.sh:
```diff
@@ -0,0 +1,28 @@
+. ${DIR_BIN_LIB}/stfw/domain/service/spec/process_spec
+
+# テスト対象の関数名 . はだめ _ は ok
+# bashの?で落ちる場合、処理続行されない（検証できない）
```
引数のnullチェック「${1:?}」は「signatureがアンマッチなコンパイルエラー」の扱い
内部で明示的にチェックしているモノ「checks.xxx」は「チェック仕様」に当たるので、テスト対象
くらいで良いかと思いますー

suwa-sh commented:
↑にプラスで、ut実行時の stfw project をどう管理するか、とか整理しないと
UTを量産するのが難しそうですねf(^_^;

レビューコメントの対応と合わせて、ちょっと整理してみますねー

suwa-sh reviewed (changes requested):
↑にプラスで、ut実行時の stfw project をどう管理するか、とか整理しないと
UTを量産するのが難しそうですねf(^_^;

レビューコメントの対応と合わせて、ちょっと整理してみますねー

tkhr-sait commented:
指摘ありがとうございます！
全体的に雑ですいませんです(^^;
1. itでの環境＋シナリオ構築あたりまでをutの前提にできれば量産しやすそうですかね。
2. 一般的な順番とは違うかもですが、
　itで大まかな動作（デグレ）確認＋環境構築後、utで詳細確認って流れがやりやすそうですかね
3. 現状、ut_allでエラートラップしてないので、全素通りですね...
4.はいー

suwa-sh committed:
fix(lib/commonns/bash_utils): コマンド組み立てが、実行になってしまっていたので修正

suwa-sh committed:
feat(lib/commons/logging): terminalで実行されている場合、ログをカラーで出力

digdagで実行される場合、ログにカラーコードが出力されてしまうので
ttyがない場合、カラーコード指定なしで出力しています。

suwa-sh committed:
fix(shunit2): カラー出力指定時に、echo に-eが付かないパターンがある

suwa-sh committed:
chore(process_spec): 命名規則を合わせた

suwa-sh committed:
test(*): UT整備

test/ut/resources/projects 配下に、構築ステップごとのtemplateを追加。
利用するときは、oneTimeSetupでディレクトリコピー、oneTimeTearDownで削除。
テストスクリプトとテストリソースを管理できる用にディレクトリ構成を変更。

suwa-sh commented:
coverageはまだですが、UT量産に向けて用意してみました。
こんなもんでどうでしょう？

1. **itでの環境＋シナリオ構築あたりまでをutの前提にできれば量産しやすそうですかね。**
template追加。oneTimeSetupで一時ディレクトリにコピー、oneTimeTearDownで削除。
これでut -> itの順番でも行けそうかな、と。

3. **test失敗時の制御を追加する必要がありそう**
考慮追加してみましたー

tkhr-sait commented:
1. 良いかと思います！
　でも仕様変更の場合にテンプレートの修正を忘れてしまうとデグレに
　気がつくのが遅れそうですね...そこはitでカバーかな？
　変則的でUTではなくなりそうですが、大まかな流れ部分はできているので
　テンプレートの生成部分もコマンドで〜っていうのもありかなと思ったり。
2.ありがとうございました！

suwa-sh commented:
1. そうですねf(^_^;
半端に共通化しないで、基本の、テスト対象に必要なものはハーネスで揃えるにしましょうかね。
他のコマンド依存も、同じ感じでややこしくなりそうなので、test/resouces/配下にファイルも手で揃える方針に。
手動で作るときは、コマンド利用OK。
テンプレ変更は、ITで落とす仕掛けを考える。
ですかねー
